openlibrary/plugins/upstream/merge_authors.py

```diff
@@ -1,77 +1,121 @@
 """Merge authors.
 """
-import web
 import re
+
 import simplejson
+import six
+import web
+
+from infogami.infobase.client import ClientException
 from infogami.utils import delegate
 from infogami.utils.view import render_template, safeint
-from infogami.infobase.client import ClientException
-
 from openlibrary.plugins.worksearch.code import top_books_from_author
 from openlibrary.utils import uniq, dicthash
 
-import six
+
+class BasicRedirectEngine:
+    """
+    Creates redirects whilst updating backrefs
+    """
+    def make_redirects(self, master, duplicates):
+        """
+        :param str master:
+        :param list of str duplicates:
+        :rtype: list of dict
+        """
+        # Create the actual redirect objects
+        docs_to_save = [make_redirect_doc(key, master) for key in duplicates]
+
+        # find the references of each duplicate and convert them
+        references = self.find_all_backreferences(duplicates)
+        docs = get_many(references)
+        docs_to_save.extend(
+            self.update_backreferences(doc, master, duplicates) for doc in docs)
+        return docs_to_save
+
+    def find_backreferences(self, key):
+        """
+        Returns keys of all the docs which have a reference to the given key.
+        All the subclasses must provide an implementation for this method.
+        :param str key: e.g. /works/OL1W
+        :rtype: list of str
+        """
+        raise NotImplementedError()
+
+    def find_all_backreferences(self, keys):
+        backrefs = set(ref for key in keys for ref in self.find_backreferences(key))
+        return list(backrefs)
+
+    def update_backreferences(self, doc, master, duplicates):
+        """
+        Converts references to any of the duplicates in the given doc to the master.
+
+        :param A doc:
+        :param str master:
+        :param list of str duplicates:
+        :rtype: A
+        """
+        if isinstance(doc, dict):
+            if len(doc) == 1 and doc.keys() == ['key']:
```

Comment:
This will never be True in Python 3.
```suggestion
            if list(doc) == ['key']:
```

Reply:
You've fixed in one of your PRs :) Rebased.